### PR TITLE
Fix #8314: Define TupleXXL equals and hashCode as in case classes

### LINKED ISSUE #8314: TupleXXL should use BoxesRunTime instead of java.util.Arrays.{deepEquals, deepHashCode}
## minimized code

dotty version: 0.22.0-RC1

```
scala> val t22_1 = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22)
val t22_1: (Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int
  , 
Int, Int, Int, Int, Int, Int, Int, Int) = (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22)

scala> val t22_2 = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22.0)                                                                                                                                                                                        
val t22_2: (Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int
  , 
Int, Int, Int, Int, Int, Int, Int, Double) = (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22.0)

scala> t22_1 == t22_2
val res0: Boolean = true

scala> val t23_1 = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23)
val t23_1: (Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int
  , 
Int, Int, Int, Int, Int, Int, Int, Int, Int) = (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23)

scala> val t23_2 = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23.0)
val t23_2: (Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int
  , 
Int, Int, Int, Int, Int, Int, Int, Int, Double) = (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23.0)

scala> t23_1 == t23_2
val res1: Boolean = false
```

## expectation

I think `t23_1 == t23_2` should return `true`.
also should not use `java.util.Arrays.deepHashCode` 

## references
- https://github.com/lampepfl/dotty/blob/26de50adf3ee0093401b0dbc5d88a4d1f6d3cb89/library/src/scala/runtime/TupleXXL.scala#L18
- http://eed3si9n.com/liberty-equality-and-boxed-primitive-types
- https://github.com/scala/scala/blob/v2.13.1/src/library/scala/runtime/BoxesRunTime.java#L117

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: library/src/scala/runtime/TupleXXL.scala [hunk 1]
```diff
@@ -1,23 +1,41 @@
...
 final class TupleXXL private (es: IArray[Object]) extends Product {
   assert(es.length > 22)
 
   def productElement(n: Int): Any = es(n)
   def productArity: Int = es.length
+  override def productPrefix: String = "TupleXXL"
+
+  override def toString: String = elems.asInstanceOf[Array[Object]].mkString("(", ",", ")")
+
+  override def hashCode: Int = {
```
Comment (smarter) [inline library/src/scala/runtime/TupleXXL.scala]: For case classes we normally delegate to `scala.runtime.ScalaRunTime._hashCode`, couldn't we do this here too ?

File: tests/run/i8314.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  assert((1, (1, 2)) == (1, (1, 2.0)))
+  assert(t4 == t5)
+
+  val t6: Tuple = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, Array(1, 2, 3))
+  val t7: Tuple = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, Array(1, 2, 3))
+  assert((1, 2, Array(3, 4)) != (1, 2, Array(3, 4)))
+  assert(t6 != t7)
+
+  val t8: Tuple = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26)
+  assert(t1 != t8)
```
Comment (OlivierBlanvillain) [inline tests/run/i8314.scala]: Could you maybe add a line to test hashCode? LGTM otherwise

File: library/src/scala/runtime/TupleXXL.scala (hunk 1 above)
Comment (nicolasstucki) [reply library/src/scala/runtime/TupleXXL.scala]: Yes, we can.

Though we do not seem to be delegating ti this method for case classes.

File: tests/run/i8314.scala (hunk 2 above)
Comment (nicolasstucki) [reply tests/run/i8314.scala]: Added `hashCode` tests

File: library/src/scala/runtime/TupleXXL.scala (hunk 1 above)
Comment (smarter) [reply library/src/scala/runtime/TupleXXL.scala]: > Though we do not seem to be delegating ti this method for case classes.

We do: https://github.com/lampepfl/dotty/blob/04042381ff99039884cd85d87e22f1463b8133cf/compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala#L284

Comment (nicolasstucki) [reply library/src/scala/runtime/TupleXXL.scala]: I mean that I saw some code generated in the hashCode, which I used as a template for the first implementation. Maybe we are optimizing by mistake something that we should not.

Comment (smarter) [reply library/src/scala/runtime/TupleXXL.scala]: If there's a primitive type field in the case class we do write it by hand.

Comment (nicolasstucki) [reply library/src/scala/runtime/TupleXXL.scala]: Then that was the case

**Pull request merged**